tutorials/poststack_nccl.py

```diff
@@ -0,0 +1,197 @@
+r"""
+Post Stack Inversion - 3D with Nvidia's NCCL
+=========================
+This tutorial is similar to poststack.py but it shows how to run PyLops-MPI in multi-GPU environment
+and have those GPUs communicate via NCCL
+"""
+
+import numpy as np
+import cupy as cp
+from scipy.signal import filtfilt
+from matplotlib import pyplot as plt
+from mpi4py import MPI
+
+from pylops.utils.wavelets import ricker
+from pylops.basicoperators import Transpose
+from pylops.avo.poststack import PoststackLinearModelling
+
+import pylops_mpi
+import pylops_mpi.utils
+import pylops_mpi.utils._nccl
+
+plt.close("all")
+rank = MPI.COMM_WORLD.Get_rank()
+# size = MPI.COMM_WORLD.Get_size()
+
+###############################################################################
+# This section is exactly the same as MPI version. We will keep using MPI for
+# transfering meta data i.e., shapes, dims, etc.
+
+# Model
+model = np.load("../testdata/avo/poststack_model.npz")
+x, z, m = model['x'][::3], model['z'], np.log(model['model'])[:, ::3]
+
+# Making m a 3D model
+ny_i = 20  # size of model in y direction for rank i
+y = np.arange(ny_i)
+m3d_i = np.tile(m[:, :, np.newaxis], (1, 1, ny_i)).transpose((2, 1, 0))
+ny_i, nx, nz = m3d_i.shape
+
+# Size of y at all ranks
+ny = MPI.COMM_WORLD.allreduce(ny_i)
+
+# Smooth model
+nsmoothy, nsmoothx, nsmoothz = 5, 30, 20
+mback3d_i = filtfilt(np.ones(nsmoothy) / float(nsmoothy), 1, m3d_i, axis=0)
+mback3d_i = filtfilt(np.ones(nsmoothx) / float(nsmoothx), 1, mback3d_i, axis=1)
+mback3d_i = filtfilt(np.ones(nsmoothz) / float(nsmoothz), 1, mback3d_i, axis=2)
+
+# Wavelet
+dt = 0.004
+t0 = np.arange(nz) * dt
+ntwav = 41
+wav = ricker(t0[:ntwav // 2 + 1], 15)[0]
+
+# Collecting all the m3d and mback3d at all ranks
+m3d = np.concatenate(MPI.COMM_WORLD.allgather(m3d_i))
+mback3d = np.concatenate(MPI.COMM_WORLD.allgather(mback3d_i))
+
+###############################################################################
+# NCCL communication can be easily initialized with
+# :py:func:`pylops_mpi.utils._nccl.initialize_nccl_comm` operator.
+# One can think of this as GPU-counterpart of MPI.COMM_WORLD
+nccl_comm = pylops_mpi.utils._nccl.initialize_nccl_comm()
+
+
+# Initialize DistributedArray with `base_comm_nccl = nccl_comm` and CuPy engine
```

Comment:
Move this to a text block (add ####)

Initialize DistributedArray -> We are now ready to initialize various :class:`pylops_mpi.DistributedArray` objects. Compared to the MPI tutorial, we need to make sure that we pass `base_comm_nccl = nccl_comm` and set CuPy as the engine.